### PR TITLE
Filter away duplicate zero-balance vacation carryover entitlements

### PR DESCRIPTION
**Pull request by apetro**
Responsive to feedback from Gareth at Friday's Service Center comms touchbase.

For whatever reason, the absence balance data from HRS includes two Vacation Carryover Balance entries. Maybe one from a University Staff perspective and one from an Academic Staff perspective? Dunno.

Sometimes, both of these balances are zero. It seems entirely clear that in this specific case, dropping the extra zero balances removes no meaning, and simplifies the user experience.

Before:

duplicate-zero-vacation-carryover-balances

After (mockup):

mock-up-removing-a-row

It's probably the case that we should go further, that when there's a non-zero vacation carryover balance and a zero vacation carryover balance, we should drop the zero balance even though there's only one such zero balance. That'd require a wee bit more confidence, since that'd be removing some data from the user experience rather than just removing duplicate data from the user experience.

First do no harm.

+ Retains one zero balance vacation carryover balance entry, so whatever meaning that's trying to convey, it continues to have the chance to convey meaning.
+ Does that retaining *per job*, so users with multiple active jobs, well, still no loss of meaning.
+ Does not drop non-zero-balance entries.

### CODE REVIEW

File: hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/VacationCarryoverBalancePredicateTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    @Test
+    public void doesNotMatchNull() {
+
+        assertFalse(predicate.apply(null));
+
+    }
+
+
+}
```
Comment (vertein) [inline hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/VacationCarryoverBalancePredicateTest.java]: Missing end of file new line

File: hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/VacationCarryoverBalancePredicateTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * Test that predicate does not match null.
+     */
+    @Test
+    public void doesNotMatchNull() {
+
+        assertFalse(predicate.apply(null));
+
+    }
+
```
Comment (vertein) [inline hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/VacationCarryoverBalancePredicateTest.java]: extra new line?

File: hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsIllegalArgumentExceptionOnNullInput() {
+        filter.filter(null);
+    }
+
+
+}
```
Comment (vertein) [inline hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilterTest.java]: Missing end of file new line

File: hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        beforeFiltering.add(anotherNonZeroBalanceVacationCarryover);
+
+        final List<AbsenceBalance> afterFiltering = filter.filter(beforeFiltering);
+
+        assertEquals(beforeFiltering, afterFiltering);
+
+    }
+
+
+
```
Comment (vertein) [inline hrs-portlets-ps-impl/src/test/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilterTest.java]: extra new lines?

**Review by vertein — APPROVED** [review]

File: hrs-portlets-ps-impl/src/main/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilter.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * Given a List of AbsenceBalances, return a List of AbsenceBalances with the same content as the input list except
+     * containing no zero-balance "Vacation Carryover Balance" entitlements other than first such zero-balance
+     * "Vacation Carryover Balance" per job. That is, for any given job, an employee should have at most one
+     * zero-balance "Vacation Carryover Balance", and that one should be the first such balance from the input list.
+     * @param input
+     * @return a filtered List<AbsenceBalance>
+     * @throws IllegalArgumentException if input is null
+     */
+    public List<AbsenceBalance> filter(final List<AbsenceBalance> input) {
```
Comment (apetro) [inline hrs-portlets-ps-impl/src/main/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilter.java:38]: CodeClimate is complaining about the "cognitive complexity" of `filter()` (scores 6, worse than the allowable 5).

I don't see how to make it structurally less complex. It's doing what it essentially needs to be doing: keeping around just enough state about what it has seen and deciding to include, exclude, and update that state as it goes.

**Comment by apetro** [comment]
Verified had desired effect in `test.my`.

**Pull request merged**